integration/experiment/power_sweep/gen_policy_recommendation.py

```diff
@@ -0,0 +1,287 @@
+    stats_out.close()
+
+
+def analyze_extracted_data(simplified_data, rtmodel, enmodel):
+    rtmodel.train([pl for pl, _, __ in simplified_data],
+            [rt for _, rt, __ in simplified_data])
+    enmodel.train([pl for pl, _, __ in simplified_data],
+            [en for _, __, en in simplified_data])
+
+
+def policy_min_energy(plrange, enmodel, rtmodel=None, pltdp=None, max_degradation=None):
+    if pltdp is None:
+        pltdp = max(plrange)
+
+    en_predictions = enmodel.batch_test(plrange)
+    if max_degradation is None:
+        # we don't need the runtime model in this case
+        besten, bestpl = min(zip(en_predictions, plrange))
+        if rtmodel:
+            bestrt = rtmodel.test(bestpl)
+        else:
+            bestrt = None
+        return bestpl, bestrt, besten
+    else:
+        rt_predictions = rtmodel.batch_test(plrange)
+        rt_at_tdp = rtmodel.test(pltdp)
+        constrained_values = [
+                (en, pl, rt)
+                for pl, rt, en in zip(plrange, rt_predictions, en_predictions)
+                if rt <= (1 + max_degradation) * rt_at_tdp]
+        besten, bestpl, bestrt = min(constrained_values)
+        return bestpl, bestrt, besten
+
+
+def main(argv):
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--path', help='path containing reports and machine.json', required=True)
+    parser.add_argument('--region_filter', help='comma-separated list of regions to include, default to include all regions', default=None)
+    parser.add_argument('--dump_prefix', help='prefix to dump statistics to, empty to not dump stats')
+    parser.add_argument('--min_pl', help='minimum power limit, default behavior is to use machine.json', default=-1, type=float)
+    parser.add_argument('--max_pl', help='maximum power limit, default behavior is to use machine.json', default=-1, type=float)
+    parser.add_argument('--tdp', help='tdp (sticker power limit), default behavior is to use machine.json', default=-1, type=float)
+    parser.add_argument('--max_degradation', help='maximum allowed runtime degradation, default is 0.1 (i.e., 10%%)', default=0.1, type=float)
+    parser.add_argument('--min_energy', help='ignore max degradation, just give the minimum energy possible', action='store_true')
+
+    args = parser.parse_args(argv[1:])
+
+    simplified_data = extract_from_reports(args.path, args.region_filter)
+    if args.dump_prefix:
+        dump_simplified_data(simplified_data, "%s.dat" % dump_prefix)
+        dump_stats_summary(simplified_data, "%s.stats" % dump_prefix)
+
+    rtmodel = CubicPowerModel()
+    enmodel = CubicPowerModel()
+
+    analyze_extracted_data(simplified_data, rtmodel, enmodel)
+
+    min_pl, tdp, max_pl = None, None, None
+    if min(args.min_pl, args.tdp, args.max_pl) < 0:
+        try:
+            machine_info = json.load(open(os.path.join(args.path, "machine.json")))
```

Comment:
We have a method for getting at the machine info.  Take a look at:
https://github.com/geopm/geopm/blob/dev/integration/experiment/monitor/gen_plot_achieved_power.py#L88

This can be refactored into the ```if __name__``` section.

Reply:
Ok, I moved a lot of the main function into the if __name__ section down below; I tried to keep the argument parsing and file input in that section, but the function calls and logic in the main function. I hope the separation makes sense.